src/compiler/checker.ts

```diff
@@ -16743,6 +16746,65 @@ namespace ts {
+            }
+        }
+
+        type voidUndefinedNullOrNeverTypeNode = Token<SyntaxKind.VoidKeyword | SyntaxKind.UndefinedKeyword | SyntaxKind.NullKeyword | SyntaxKind.NeverKeyword>;
+
+        function getEntityNameForDecoratoryMetadata(node: TypeNode): EntityName | voidUndefinedNullOrNeverTypeNode {
+            if (node) {
+                switch (node.kind) {
+                    case SyntaxKind.IntersectionType:
+                    case SyntaxKind.UnionType:
+                        let commonEntityName: EntityName | voidUndefinedNullOrNeverTypeNode;
+                        for (const typeNode of (<UnionOrIntersectionTypeNode>node).types) {
+                            const individualEntityName = getEntityNameForDecoratoryMetadata(typeNode);
+                            if (!individualEntityName) {
+                                // Individual is something like string number 
+                                // So it would be serialized to either that type or object 
+                                // Safe to return here
+                                return undefined;
+                            }
+
+                            const isCommonEntityName = commonEntityName && isEntityName(commonEntityName);
+                            const isIndividualEntityName = isEntityName(individualEntityName);
+                            if (isCommonEntityName && isIndividualEntityName) {
+                                // Note this is in sync with the transformation that happens for type node.
+                                // Keep this in sync with serializeUnionOrIntersectionType
+                                // Verify if they refer to same entity and is identifier
+                                // return undefined if they dont match because we would emit object
+                                if (!isIdentifier(commonEntityName) ||
+                                    !isIdentifier(individualEntityName) ||
+                                    commonEntityName.text !== individualEntityName.text) {
+                                    return undefined;
+                                }
+                            }
+                            else if (!isCommonEntityName) {
+                                commonEntityName = individualEntityName;
+                            }
+                        }
+                        return commonEntityName;
+
+                    case SyntaxKind.ParenthesizedType:
+                        return getEntityNameForDecoratoryMetadata((<ParenthesizedTypeNode>node).type);
+
+                    case SyntaxKind.TypeReference:
+                        return (<TypeReferenceNode>node).typeName;
+
+                    case SyntaxKind.VoidKeyword:
+                    case SyntaxKind.UndefinedKeyword:
+                    case SyntaxKind.NullKeyword:
+                    case SyntaxKind.NeverKeyword:
+                        return <voidUndefinedNullOrNeverTypeNode>node;

```

Comment:
Why do we return the token here? This function is only used by `markDecoratorMedataDataTypeNodeAsReferenced`, which in turn only cares if its an `EntityName`.